packages/ckeditor5-table/tests/converters/downcast.js

```diff
@@ -5,571 +5,512 @@
+				);
 			} );
 
-			assertEqualMarkup( getViewData( view, { withoutSelection: true } ), viewTable( [
-				[ '00', '01' ],
-				[ '', '' ]
-			] ) );
+			describe( 'headingColumns attribute', () => {
+				it( 'should mark heading columns table cells', () => {
+					setModelData( model, modelTable( [
+						[ '00', '01', '02' ],
+						[ '10', '11', '12' ]
+					], { headingColumns: 2 } ) );
+
+					assertEqualMarkup( editor.getData(),
+						'<figure class="table">' +
+							'<table>' +
+								'<tbody>' +
+									'<tr><th>00</th><th>01</th><td>02</td></tr>' +
+									'<tr><th>10</th><th>11</th><td>12</td></tr>' +
+								'</tbody>' +
+							'</table>' +
+						'</figure>'
+					);
+				} );
+
+				it( 'should mark heading columns table cells when one has colspan attribute', () => {
+					setModelData( model, modelTable( [
+						[ '00', '01', '02', '03' ],
+						[ { colspan: 2, contents: '10' }, '12', '13' ]
+					], { headingColumns: 3 } ) );
+
+					assertEqualMarkup( editor.getData(),
+						'<figure class="table">' +
+							'<table>' +
+								'<tbody>' +
+									'<tr><th>00</th><th>01</th><th>02</th><td>03</td></tr>' +
+									'<tr><th colspan="2">10</th><th>12</th><td>13</td></tr>' +
+								'</tbody>' +
+							'</table>' +
+						'</figure>'
+					);
+				} );
+
+				it( 'should work with colspan and rowspan attributes on table cells', () => {
+					// The table in this test looks like a table below:
+					//
+					//   Row headings | Normal cells
+					//                |
+					// +----+----+----+----+
+					// | 00 | 01 | 02 | 03 |
+					// |    +----+    +----+
+					// |    | 11 |    | 13 |
+					// |----+----+    +----+
+					// | 20      |    | 23 |
+					// |         +----+----+
+					// |         | 32 | 33 |
+					// +----+----+----+----+
+					setModelData( model, modelTable( [
+						[ { rowspan: 2, contents: '00' }, '01', { rowspan: 3, contents: '02' }, '03' ],
+						[ '11', '13' ],
+						[ { colspan: 2, rowspan: 2, contents: '20' }, '23' ],
+						[ '32', '33' ]
+					], { headingColumns: 3 } ) );
+
+					assertEqualMarkup( editor.getData(),
+						'<figure class="table">' +
+							'<table>' +
+								'<tbody>' +
+									'<tr><th rowspan="2">00</th><th>01</th><th rowspan="3">02</th><td>03</td></tr>' +
+									'<tr><th>11</th><td>13</td></tr>' +
+									'<tr><th rowspan="2" colspan="2">20</th><td>23</td></tr>' +
+									'<tr><th>32</th><td>33</td></tr>' +
+								'</tbody>' +
+							'</table>' +
+						'</figure>'
+					);
+				} );
+			} );
+
+			it( 'should create table with tbody', () => {
+				setModelData( model, modelTable( [ [ '' ] ] ) );
+
+				assertEqualMarkup( editor.getData(),
+					'<figure class="table">' +
+					'<table>' +
+					'<tbody>' +
+					'<tr><td>&nbsp;</td></tr>' +
+					'</tbody>' +
+					'</table>' +
+					'</figure>'
+				);
+			} );
 		} );
+	} );
+
+	describe( 'downcastInsertRow()', () => {
+		// The insert row downcast conversion is not executed in data pipeline.
```

Comment:
Maybe we should add test to make sure that row conversion is not used in data pipeline?

Reply:
I'm not able to think of such test :) It would be very synthetic. The data pipeline conversion runs on whole, static content it do not convert "dynamic" actions (adding/removing stuff, changing attributes, etc).